### PR TITLE
fix(table): removed first/last child padding, added inset

### LINKED ISSUE #7032: Table first/last child padding updates
TLDR;

Table `:first/:last-child` css declarations are relics that should be standardized. As per a recent conversation w/ @lboehling, `:first/:last-child` `th/td` padding definitions can/should be simplified, while inset updates should be optional.

Why this is important:
- `:first/:last-child` are detached `:pseudo` elements and are calculated by the browser on every page
- These selectors cause a `force-reflow`, which significantly impacts LCP ([Largest Contentful Paint](https://developer.chrome.com/docs/lighthouse/performance/lighthouse-largest-contentful-paint/?utm_source=lighthouse&utm_medium=devtools)), which is the time the largest content element in the viewport is rendered to the screen.
- It's unnecessary 

This update includes
- [ ] Remove default `:first/:last-child` padding modifications
- [ ] Add optional `pf-m-inset-[sm, md, lg, xl]` `:first/:last-child` padding modifier

[Backstop report](https://drive.google.com/file/d/1dimvgl3HCcT0TIZC9olhuACTWPpvNA3z/view?usp=sharing)

### PR DESCRIPTION
**Pull request by mattnolting**
closes #7032  
closes [#11058](https://github.com/patternfly/patternfly/issues/7112)

Updated [Table Iso Backstop Report](https://drive.google.com/file/d/1aj7jx_rWbyy5eCK0738ZdcKh_AGa_bBx/view?usp=sharing)
Updated [Full Backstop Report](https://drive.google.com/file/d/1NKGM1Q2wz6V3Buyvj4w4Cx6BKRAuOf6R/view?usp=sharing)

Examples: https://patternfly-pr-7050.surge.sh/components/table

# TLDR:
Floating `:first/:last-child` pseudos result in a huge number of selector matching attempts. Attaching these pseudos to a selector results in a massive reduction in matching attempts. 

The cell padding and button fitting updates are the result of a somewhat hidden bug. `thead` row heights are affected erratically by using negative margining on `table__button` cells. See below:

**Before**

Screenshot 2024-09-17 at 10 56 47 AM

Screenshot 2024-09-17 at 10 56 05 AM

Screenshot 2024-09-17 at 10 53 34 AM

Screenshot 2024-09-17 at 10 52 39 AM

**After**

Screenshot 2024-09-17 at 10 57 35 AM

Screenshot 2024-09-17 at 10 57 23 AM
Screenshot 2024-09-17 at 10 57 10 AM
Screenshot 2024-09-17 at 10 56 58 AM

~~Table :first/:last-child css declarations are relics that should be standardized. As per a recent conversation w/ @lboehling, :first/:last-child th/td padding definitions can/should be simplified, while inset updates should be optional.~~

~~Also, toggles, checkboxes, radios, actions, etc do not need special spacing.~~

These need to be updated

## Performance impact

This PR 
* Reduces match attempts by **4,361,984**
* Improves performance by 2%
* Reduces blocking time by 50ms

**Before**

Screenshot 2024-09-10 at 11 57 31 AM

**After**

Screenshot 2024-09-10 at 11 49 02 AM

**Before**

Screenshot 2024-09-10 at 12 03 46 PM

**After**

Screenshot 2024-09-10 at 12 05 26 PM

[Backstop report](https://drive.google.com/file/d/1XSd0422s0tgllaC3dVdIir6kcamoFDiU/view)

### CODE REVIEW

**Review by mcoker — COMMENTED** [review]
I'm seeing slightly different results in the backstop report - [backstop-first-last-child-table.pdf](https://github.com/user-attachments/files/16982214/backstop-first-last-child-table.pdf).

At a glance, it looks like expandable toggles are no longer aligned with the row text? https://patternfly-pr-7050.surge.sh/components/table#expandable

cc @lboehling, would you mind looking at the report or table examples and see if this looks as you'd expect?

**Comment by mattnolting** [comment]
# TLDR:

Line height has been standardized to create consistent row height. Currently, rows without actionable elements are shorter than rows with actionable elements. Table sort buttons are shorter than plain buttons (row toggles) which creates a presentational difference, especially notable when a `thead toggle` button is adjacent to a `thead sort` button. Because `line-height` is defined as `1.5`, `thead` block height resolves to `18px`, instead of `21px`, which is the default for `button`s or `menu-toggle`s. It also results in vertical misalignment as some elements may that may not be as tall as others. In order to correct this, a standardized `line-height` of `table--LineHeight * table--FontSize`(not `th/td LineHeight * th/td FontSize)` has been applied to all cells. The result is a consistent block height that is independent of `font-size`. 

@lboehling @kaylachumley @andrew-ronaldson Can you take a look at [Backstop report](https://drive.google.com/file/d/1dimvgl3HCcT0TIZC9olhuACTWPpvNA3z/view)? 

## What changed

- [caption `inline-padding`](https://github.com/patternfly/patternfly/pull/7050/files#diff-393474b82acf1f2bc6bca67b97880a38b52ef4a23ff35da5875e6ce06afedb39R88-R91) has been updated to match `th/td` `inline-padding`
- [`line-height` has been standardized](https://github.com/patternfly/patternfly/pull/7050/files#diff-393474b82acf1f2bc6bca67b97880a38b52ef4a23ff35da5875e6ce06afedb39R215-R216) to create consistent row height.
- First/last child table cells have been [standardized to match `menu-toggle` padding](https://github.com/patternfly/patternfly/pull/7050/files#diff-393474b82acf1f2bc6bca67b97880a38b52ef4a23ff35da5875e6ce06afedb39R44)
- [`radio/checkbox` alignment has been corrected](https://github.com/patternfly/patternfly/pull/7050/files#diff-393474b82acf1f2bc6bca67b97880a38b52ef4a23ff35da5875e6ce06afedb39R729)

**Help**

Screenshot 2024-09-20 at 3 46 46 PM

Screenshot 2024-09-20 at 3 47 43 PM

**Checkboxes**

Screenshot 2024-09-20 at 3 16 22 PM

Screenshot 2024-09-20 at 3 16 31 PM

**Radio**

Screenshot 2024-09-20 at 3 33 18 PM

**Expandable**

Screenshot 2024-09-20 at 3 17 46 PM

Screenshot 2024-09-20 at 3 17 35 PM

**Compact**

Screenshot 2024-09-20 at 3 30 26 PM

Screenshot 2024-09-20 at 3 30 32 PM


**Compound expansion**

Screenshot 2024-09-20 at 3 34 39 PM
 
Screenshot 2024-09-20 at 3 27 25 PM

**Favorites**

Screenshot 2024-09-20 at 3 28 42 PM

**Compact**

Screenshot 2024-09-20 at 3 30 26 PM

Screenshot 2024-09-20 at 3 30 32 PM

**Borderless**

Screenshot 2024-09-20 at 3 28 12 PM

**Compound expansion**

**Tree table**

Screenshot 2024-09-20 at 3 28 01 PM

Screenshot 2024-09-20 at 3 27 48 PM

- Rather than negative margins on sort buttons, `table__action` cells have been standardized to exclude top padding (in `thead` cells) and bottom padding (in `tbody` cells). Margin top has been added to `thead` sort buttons and bottom margin has been added to `tbody` sort buttons to retain a common default height of 53px when standard `th/td` content is absent. 

Screenshot 2024-09-20 at 3 39 15 PM

Screenshot 2024-09-20 at 3 39 20 PM

- Specific padding variables `__toggle`, `__check`, `__action` , `__draggable`, `__favorite` [have been abstracted](https://github.com/patternfly/patternfly/pull/7050/files#diff-393474b82acf1f2bc6bca67b97880a38b52ef4a23ff35da5875e6ce06afedb39R48-R51) to `--#{$table}--cell--action--PaddingBlockStart/End: var(--pf-t--global--spacer--control--vertical--plain);` and `--#{$table}--cell--action--PaddingInlineStart/End: var(--pf-t--global--spacer--action--horizontal--plain--default);`

Screenshot 2024-09-20 at 3 40 12 PM

## Mobile presentation has been simplified and has changed slightly as a result

- `padding-block-start/end` is consistent

Screenshot 2024-09-20 at 4 00 29 PM

Screenshot 2024-09-20 at 4 01 23 PM

Screenshot 2024-09-20 at 4 01 58 PM

Screenshot 2024-09-20 at 4 02 27 PM

Screenshot 2024-09-20 at 4 02 53 PM

File: src/patternfly/components/Table/table-tr.hbs [hunk 1]
```diff
@@ -23,8 +23,8 @@
       {{ternary table-tr--IsHidden 'hidden' null}}
     {{/unless}}
   {{/if}}
-  {{ternary table--IsGrid 'role="row"'}}
-  {{ternary table-tr--IsClickable 'tabindex="0"'}}
+  {{ternary table--IsGrid 'role="row"' null}}
+  {{ternary table-tr--IsClickable 'tabindex="0"' null}}
```
Comment (srambach) [inline src/patternfly/components/Table/table-tr.hbs:27]: I'm seeing <tr class="pf-v6-c-table__tr" role&#x26;#x3d;&#x26;quot;row&#x26;quot;=""> so I think something is amiss here. (e.g. on http://localhost:5000/components/table/html/basic-table-example/)

**Review by srambach — COMMENTED** [review]
I haven't made it all the way through, but ran regression tests on the table examples. I'm seeing some shifts in all different directions and really can't discern what's correct. 

I believe the line height change causes some to move up, some down. Spot checks on the desktop width show they are 21px with 16px top and bottom, so that's good.

Mobile view seems to show increased height; is that also intended? e.g. pf-core__components_table_html_sortable-example_0_document_0_mobile.png

Left-right shifts seem to be a difference in padding - I can't actually tell from the issue description - what's the desired padding?
 
A couple of other things I noticed:
 /components/table/html/borderless-expandable-example has some expanded content that lost it's left padding

/components/table/html/borderless-with-compound-expansion-example  the kebabs no longer align horizontally between the main table and expanded section

File: src/patternfly/components/Table/table-tr.hbs (hunk 1 above)
Comment (mattnolting) [reply src/patternfly/components/Table/table-tr.hbs]: This issue was pre-existing and is now resolved 👍

File: src/patternfly/components/Button/button.scss [hunk 2]
```diff
@@ -5,11 +5,11 @@
   --#{$button}--AlignItems: baseline;
   --#{$button}--JustifyContent: center;
   --#{$button}--Gap: var(--pf-t--global--spacer--gap--text-to-element--default);
-  --#{$button}--MinWidth: calc(1em * var(--#{$button}--LineHeight) + var(--#{$button}--PaddingBlockStart) + var(--#{$button}--PaddingBlockEnd));
   --#{$button}--PaddingBlockStart: var(--pf-t--global--spacer--control--vertical--default);
-  --#{$button}--PaddingInlineEnd: var(--pf-t--global--spacer--action--horizontal--default);
   --#{$button}--PaddingBlockEnd: var(--pf-t--global--spacer--control--vertical--default);
+  --#{$button}--PaddingInlineEnd: var(--pf-t--global--spacer--action--horizontal--default);
   --#{$button}--PaddingInlineStart: var(--pf-t--global--spacer--action--horizontal--default);
+  --#{$button}--MinWidth: calc(1em * var(--#{$button}--LineHeight) + var(--#{$button}--PaddingBlockStart) + var(--#{$button}--PaddingBlockEnd));
```
[the remark this answers is not in the document]
Comment (mattnolting) [reply src/patternfly/components/Button/button.scss]: Good point. This shouldn't have been in this PR. I've also reverted any other, non-necessary CSS/CSS var updates.

**Comment by mcoker** [comment]
Backstop report PDF - https://drive.google.com/file/d/1KK0zfG0OJPquXKql5AhnLY8pKRWMJWL-/view?usp=sharing
Full report - https://drive.google.com/file/d/13jjqKLZjH9kVlSKXFDVRui9C7GLVRbQu/view?usp=sharing
* You'll need to download, unzip, then open `backstop_data -> html_report -> index.html`

File: src/patternfly/components/Table/table-grid.scss [hunk 3]
```diff
@@ -98,17 +98,19 @@
...
   --#{$table}--m-compact__tr--responsive--PaddingBlockEnd: var(--pf-t--global--spacer--sm);
-  --#{$table}--m-compact__tr__td--responsive--PaddingBlockStart: var(--pf-t--global--spacer--xs);
-  --#{$table}--m-compact__tr__td--responsive--PaddingBlockEnd: var(--pf-t--global--spacer--xs);
   --#{$table}--m-compact__check--responsive--MarginBlockStart: #{pf-size-prem(7px)};
   --#{$table}--m-compact__action--responsive--MarginBlockStart: calc(var(--pf-t--global--spacer--xs) * -1);
   --#{$table}--m-compact__action--responsive--MarginBlockEnd: calc(var(--pf-t--global--spacer--xs) * -1);
   --#{$table}--m-compact__toggle--c-button--responsive--MarginBlockEnd: calc(var(--pf-t--global--spacer--xs) * -1);
 
+  --#{$table}--m-compact--cell--responsive--PaddingBlockStart: var(--pf-t--global--spacer--sm);
+  --#{$table}--m-compact--cell--responsive--PaddingBlockEnd: var(--pf-t--global--spacer--sm);
```
Comment (srambach) [inline src/patternfly/components/Table/table-grid.scss:107]: These two variables don't seem to be used?

**Comment by mattnolting** [comment]
@mcoker @srambach @andrew-ronaldson @lboehling @kaylachumley [Updated Backstop report](https://drive.google.com/file/d/1aj7jx_rWbyy5eCK0738ZdcKh_AGa_bBx/view?usp=sharing)

**Comment by mattnolting** [comment]
Updated [Table Iso Backstop Report](https://drive.google.com/file/d/1aj7jx_rWbyy5eCK0738ZdcKh_AGa_bBx/view?usp=sharing)

File: src/patternfly/components/Table/table-grid.scss [hunk 4]
```diff
@@ -78,50 +79,53 @@
...
 
   // * Table grid compact table
   --#{$table}--m-compact__tr--responsive--PaddingBlockStart: var(--pf-t--global--spacer--sm);
   --#{$table}--m-compact__tr--responsive--PaddingBlockEnd: var(--pf-t--global--spacer--sm);
-  --#{$table}--m-compact__tr__td--responsive--PaddingBlockStart: var(--pf-t--global--spacer--xs);
-  --#{$table}--m-compact__tr__td--responsive--PaddingBlockEnd: var(--pf-t--global--spacer--xs);
-  --#{$table}--m-compact__check--responsive--MarginBlockStart: #{pf-size-prem(7px)};
-  --#{$table}--m-compact__action--responsive--MarginBlockStart: calc(var(--pf-t--global--spacer--xs) * -1);
-  --#{$table}--m-compact__action--responsive--MarginBlockEnd: calc(var(--pf-t--global--spacer--xs) * -1);
+  --#{$table}--m-compact__check--responsive--MarginBlockStart: var(--pf-t--global--spacer--sm);
```
Comment (mattnolting) [inline src/patternfly/components/Table/table-grid.scss:105]: This can be removed as it's no longer used.

File: src/patternfly/components/Table/table-grid.scss [hunk 5]
```diff
@@ -78,50 +79,53 @@
...
+  --#{$table}--m-grid--cell--PaddingBlockEnd: var(--pf-t--global--spacer--md);
   --#{$table}--m-grid--cell--PaddingInlineEnd: 0;
-  --#{$table}--m-grid--cell--PaddingBlockEnd: 0;
   --#{$table}--m-grid--cell--PaddingInlineStart: 0;
 
-  // * Table td responsive
-  --#{$table}-td--responsive--GridColumnGap: var(--pf-t--global--spacer--md);
-  --#{$table}--cell--responsive--PaddingBlockStart: var(--pf-t--global--spacer--md);
-  --#{$table}--cell--responsive--PaddingBlockEnd: var(--pf-t--global--spacer--sm);
-  --#{$table}--cell--first-child--responsive--PaddingBlockStart: var(--pf-t--global--spacer--sm);
```
Comment (mattnolting) [inline src/patternfly/components/Table/table-grid.scss:94]: No longer necessary.

File: src/patternfly/components/Table/table-grid.scss [hunk 6]
```diff
@@ -130,28 +134,65 @@
...
-  --#{$table}--cell--PaddingBlockStart: var(--#{$table}--m-grid--cell--PaddingBlockStart);
-  --#{$table}--cell--PaddingInlineEnd: var(--#{$table}--m-grid--cell--PaddingInlineEnd);
-  --#{$table}--cell--PaddingBlockEnd: var(--#{$table}--m-grid--cell--PaddingBlockEnd);
-  --#{$table}--cell--PaddingInlineStart: var(--#{$table}--m-grid--cell--PaddingInlineStart);
-
-  // - Table favorite
-  --#{$table}__favorite--c-button--MarginBlockStart: auto;
-  --#{$table}__favorite--c-button--MarginInlineEnd: auto;
-  --#{$table}__favorite--c-button--MarginBlockEnd: auto;
-  --#{$table}__favorite--c-button--MarginInlineStart: auto;
```
Comment (mattnolting) [inline src/patternfly/components/Table/table-grid.scss:142]: `.#{$table}__favorite` is part of the standardized, "action item" group. Specific `margin` settings are no longer necessary.

**Comment by mattnolting** [comment]
[Full Backstop Report](https://drive.google.com/file/d/1NKGM1Q2wz6V3Buyvj4w4Cx6BKRAuOf6R/view?usp=sharing)

File: src/patternfly/components/Table/table.scss [hunk 7]
```diff
@@ -715,20 +724,18 @@
...
   --#{$table}--cell--MaxWidth: auto;
 }
 
 // - Table toggle
 .#{$table}__toggle {
-  --#{$table}--cell--PaddingBlockStart: var(--#{$table}__toggle--PaddingBlockStart);
-  --#{$table}--cell--PaddingBlockEnd: var(--#{$table}__toggle--PaddingBlockEnd);
-  --#{$table}--cell--PaddingInlineStart: var(--#{$table}__toggle--PaddingInlineStart);
-  --#{$table}--cell--PaddingInlineEnd: var(--#{$table}__toggle--PaddingInlineEnd);
-
```
Comment (mattnolting) [inline src/patternfly/components/Table/table.scss:731]: Toggle no longer needs a specific padding as it belongs to the `.#{$table}--cell--action` group.

File: src/patternfly/components/Table/table.scss [hunk 8]
```diff
@@ -750,33 +757,9 @@
...
-  }
-
-  .#{$check}.pf-m-standalone,
-  .#{$radio}.pf-m-standalone {
-    thead & {
-      display: table-cell;
-      height: auto;
-      line-height: 1;
-      vertical-align: baseline;
-    }
```
Comment (mattnolting) [inline src/patternfly/components/Table/table.scss:779]: Special styling for `__check` is no longer necessary as its height has been standardized to `table-_FontSize * table__LineHeight` or `21px` resulting in vertical alignment with all other elements.

File: src/patternfly/components/Table/table.scss [hunk 9]
```diff
@@ -806,15 +789,28 @@
...
-  --#{$table}--cell--PaddingInlineStart: var(--#{$table}__action--PaddingInlineStart);
-  --#{$table}--cell--PaddingInlineEnd: var(--#{$table}__action--PaddingInlineEnd);
+.#{$table}__draggable,
+.#{$table}__sort,
+.#{$table}__toggle {
+  // set tbody actionable cells padding-block-start to zero, as negative margins impact row height
+  .#{$table}__thead & {
+    padding-block-start: 0;
+    padding-block-end: var(--#{$table}--cell--action--PaddingBlockEnd);
+  }
```
Comment (mattnolting) [inline src/patternfly/components/Table/table.scss:803]: For `thead` `.#{$table}--cell--action` cells, only bottom padding needs to be applied in order to create consistent vertical alignment, as `thead` is set to `vertical-align: bottom`.

File: src/patternfly/components/Table/table.scss [hunk 10]
```diff
@@ -806,15 +789,28 @@
...
+
+  // set tbody actionable cells padding-block-start to zero, as negative margins impact row height
+  .#{$table}__tbody & {
+    padding-block-start: var(--#{$table}--cell--action--PaddingBlockStart);
+    padding-block-end: 0;
+  }
```
Comment (mattnolting) [inline src/patternfly/components/Table/table.scss:809]: For `tbody` `.#{$table}--cell--action` cells, only bottom padding needs to be applied in order to create consistent vertical alignment, as `tbody` is set to `vertical-align: top`.
